ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java

```diff
@@ -134,15 +134,20 @@ public AmbryRequests(StoreManager storeManager, RequestResponseChannel requestRe
     this.connectionPool = connectionPool;
     this.metricRegistry = registry;
     this.serverConfig = serverConfig;
-    transformer = null;
-    if (serverConfig != null) {
-      try {
-        StoreKeyConverter keyConverter = storeKeyConverterFactory.getStoreKeyConverter();
-        transformer = Utils.getObj(serverConfig.serverMessageTransformer, storeKeyFactory, keyConverter);
-      } catch (Exception e) {
-        logger.error("Failed to create transformer", e);
+    /* All the request handlers share one single AmbryRequests object.
+     * But the StoreKeyConverter of the Transformer has a cache which shouldn't be shared among handlers.
+     * So use ThreadLocal transformer. */
+    this.transformer = ThreadLocal.withInitial(() -> {
+      if (serverConfig != null) {
```

Comment:
can we just require serverConfig to be not null?

Reply:
It's very possible there is some legacy code or test code doesn't provide it. So I added this check.
Our production code always provide it.